flink-clients/src/main/java/org/apache/flink/client/program/PackagedProgramRetrieverImpl.java

```diff
@@ -0,0 +1,203 @@
+import org.apache.flink.client.deployment.application.EntryClassInformationProvider;
+import org.apache.flink.client.deployment.application.FromClasspathEntryClassInformationProvider;
+import org.apache.flink.client.deployment.application.FromJarEntryClassInformationProvider;
+import org.apache.flink.util.FileUtils;
+import org.apache.flink.util.FlinkException;
+import org.apache.flink.util.Preconditions;
+import org.apache.flink.util.function.FunctionUtils;
+
+import javax.annotation.Nullable;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/** {@code PackageProgramRetrieverImpl} implements {@link PackagedProgramRetriever}. */
+public class PackagedProgramRetrieverImpl implements PackagedProgramRetriever {
+
+    private final EntryClassInformationProvider entryClassInformationProvider;
+    private final String[] programArguments;
+    private final List<URL> userClasspath;
+
+    /**
+     * Creates a {@code PackageProgramRetrieverImpl} with the given parameters.
+     *
+     * @param userLibDir The user library directory that is used for generating the user classpath
+     *     if specified. The system classpath is used if not specified.
+     * @param jobClassName The job class that will be used if specified. The classpath is used to
+     *     detect any main class if not specified.
+     * @param programArgs The program arguments.
+     * @return The {@code PackageProgramRetrieverImpl} that can be used to create a {@link
+     *     PackagedProgram} instance.
+     * @throws FlinkException If something goes wrong during instantiation.
+     */
+    public static PackagedProgramRetrieverImpl create(
+            @Nullable File userLibDir, @Nullable String jobClassName, String[] programArgs)
+            throws FlinkException {
+        return create(userLibDir, null, jobClassName, programArgs);
+    }
+
+    /**
+     * Creates a {@code PackageProgramRetrieverImpl} with the given parameters.
+     *
+     * @param userLibDir The user library directory that is used for generating the user classpath
+     *     if specified. The system classpath is used if not specified.
+     * @param jarFile The jar archive used expecting to have the job class included.
+     * @param jobClassName The job class that will be used if specified. The classpath is used to
+     *     detect any main class if not specified.
+     * @param programArgs The program arguments.
+     * @return The {@code PackageProgramRetrieverImpl} that can be used to create a {@link
+     *     PackagedProgram} instance.
+     * @throws FlinkException If something goes wrong during instantiation.
+     */
+    public static PackagedProgramRetrieverImpl create(
+            @Nullable File userLibDir,
+            @Nullable File jarFile,
+            @Nullable String jobClassName,
+            String[] programArgs)
```

Comment:
Does this need to add `Configuration` parameter?

Reply:
Thanks for being so responsive looking into the PR. You're right. The draft is still a draft. I'm gonna add the fixes for configuration and savepoint settings soon. I needed to create the draft first because I wanted to link it to [FLINK-23154](https://issues.apache.org/jira/browse/FLINK-23154).